src/kit/taosdemo/taosdemo.c

```diff
@@ -9616,217 +9651,459 @@ static void* syncWriteInterlaceStmt(threadInfo *pThreadInfo, uint32_t interlaceR
+                    return -1;
                 }
-                rand_string(buf, stbInfo->columns[c].dataLen);
-                if (stbInfo->columns[c].data_type ==
-                    TSDB_DATA_TYPE_BINARY) {
-                    dataLen += snprintf(line + dataLen,
-                                        BUFFER_SIZE - dataLen,
-                                        "c%d=\"%s\"", c, buf);
+                rand_string(buf, stbInfo->tags[i].dataLen);
+                if (stbInfo->tags[i].data_type == TSDB_DATA_TYPE_BINARY) {
+                    cJSON_AddStringToObject(tag, "value", buf);
+                    cJSON_AddStringToObject(tag, "type", "binary");
                 } else {
-                    dataLen += snprintf(line + dataLen,
-                                        BUFFER_SIZE - dataLen,
-                                        "c%d=L\"%s\"", c, buf);
+                    cJSON_AddStringToObject(tag, "value", buf);
+                    cJSON_AddStringToObject(tag, "type", "nchar");    
                 }
-                tmfree(buf);
                 break;
             default:
-                errorPrint2("%s() LN%d, Unknown data type %s\n",
-                            __func__, __LINE__,
-                            stbInfo->columns[c].dataType);
-                exit(EXIT_FAILURE);
+                break;
         }
+    cJSON_AddItemToObject(tags, tagName, tag);
     }
-    dataLen += snprintf(line + dataLen, BUFFER_SIZE - dataLen," %" PRId64 "", timestamp);
+    tmfree(tagName);
+    tmfree(tbName);
+    return 0;
+}
+
+static int32_t generateSmlJsonCols(cJSON* array, cJSON* tag, SSuperTable* stbInfo,
+                                threadInfo* pThreadInfo, int64_t timestamp) {
+    cJSON* record = cJSON_CreateObject();
+    cJSON_AddStringToObject(record, "metric", stbInfo->stbName);
+    cJSON* ts = cJSON_CreateObject();
+    cJSON_AddNumberToObject(ts, "value", timestamp);
+    if (pThreadInfo->time_precision == TSDB_TIME_PRECISION_MILLI) {
+        cJSON_AddStringToObject(ts, "type", "ms");
+    } else if (pThreadInfo->time_precision == TSDB_TIME_PRECISION_MICRO) {
+        cJSON_AddStringToObject(ts, "type", "us");
+    } else if (pThreadInfo->time_precision == TSDB_TIME_PRECISION_NANO) {
+        cJSON_AddStringToObject(ts, "type", "ns");
+    } else {
+        errorPrint2("[%d]: %s() LN%d, unsupport time precision %d\n",
+            pThreadInfo->threadID, __func__, __LINE__, pThreadInfo->time_precision);
+        return -1;
+    }
+    cJSON_AddItemToObject(record, "timestamp", ts);
+    cJSON* value = cJSON_CreateObject();
+    switch (stbInfo->columns[0].data_type) {
+        case TSDB_DATA_TYPE_BOOL:
+            cJSON_AddNumberToObject(value, "value", rand_bool());
+            cJSON_AddStringToObject(value, "type", "bool");
+            break;
+        case TSDB_DATA_TYPE_UTINYINT:
+        case TSDB_DATA_TYPE_TINYINT:
+            cJSON_AddNumberToObject(value, "value", rand_tinyint());
```

Comment:
should we support rand_utinyint(), rand_usmallint(), etc to test boundary more precisely?

Reply:
We don't support unsigned data type for schemaless JSON protocol, I can either change the unsigned data to invalid data type or keep this which deals with unsigned and signed data in the same way.

Reply:
let user know unsigned is invalid is better.